nuxeo-features/nuxeo-platform-tag-service/nuxeo-platform-tag-core/src/main/java/org/nuxeo/ecm/platform/tag/FacetedTagService.java

```diff
@@ -0,0 +1,153 @@
+                    if (tags.contains(label)) {
+                        tags.remove(label);
+                    }
+                }
+                setTags(docModel, tags);
+                session.saveDocument(docModel);
+            }
+        }
+    }
+
+    @Override
+    public Set<String> doGetDocumentTags(CoreSession session, String docId) {
+        DocumentModel docModel = session.getDocument(new IdRef(docId));
+        if (!docModel.hasFacet(TAG_FACET)) {
+            return Collections.emptySet();
+        }
+        return new HashSet<>(getTags(docModel));
+    }
+
+    @Override
+    public void doCopyTags(CoreSession session, String srcDocId, String dstDocId, boolean removeExistingTags) {
+        DocumentModel srcDocModel = session.getDocument(new IdRef(srcDocId));
+        DocumentModel dstDocModel = session.getDocument(new IdRef(dstDocId));
+
+        if (!dstDocModel.isProxy()) {
+            List<String> srcTags = getTags(srcDocModel);
+            List<String> dstTags;
+            if (removeExistingTags) {
+                dstTags = srcTags;
+            } else {
+                dstTags = getTags(dstDocModel);
+                for (String tag : srcTags) {
+                    if (!dstTags.contains(tag)) {
+                        dstTags.add(tag);
+                    }
+                }
+            }
+            setTags(dstDocModel, dstTags);
+            // Disable auto checkout of destination document model
+            // to disable deletion of a version which could be the base of the live document
+            if (srcDocModel.isVersion()) {
+                dstDocModel.putContextData(VersioningService.DISABLE_AUTO_CHECKOUT, Boolean.TRUE);
+            }
+            session.saveDocument(dstDocModel);
+        }
+    }
+
+    @Override
+    public List<String> doGetTagDocumentIds(CoreSession session, String label) {
+        List<Map<String, Serializable>> res = getItems(PAGE_PROVIDERS.GET_DOCUMENT_IDS_FOR_FACETED_TAG.name(), session,
+                label);
+        if (res == null) {
+            return Collections.emptyList();
+        }
+        return res.stream().map(m -> (String) m.get(ECM_UUID)).collect(Collectors.toList());
+    }
+
+    @Override
+    public Set<String> doGetTagSuggestions(CoreSession session, String label) {
+        List<Map<String, Serializable>> res = getItems(PAGE_PROVIDERS.GET_FACETED_TAG_SUGGESTIONS.name(), session,
+                label);
+        if (res == null) {
+            return Collections.emptySet();
+        }
+        return res.stream().map(m -> (String) m.get(TagConstants.TAG_LIST + "/*1")).collect(Collectors.toSet());
+    }
+
+    @Override
+    public List<Tag> getTagCloud(CoreSession session, String docId, String username, Boolean normalize) {
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    protected List<String> getTags(DocumentModel docModel) {
+        return (List<String>) docModel.getPropertyValue(TAG_LIST);
```

Comment:
You should catch `PropertyNotFoundException` and return `Collections.emptyList()` in that case.